src/v/datalake/values_parquet.cc

```diff
@@ -0,0 +1,142 @@
+        auto high_part = ss::cpu_to_be(Int128High64(v.val));
+        buffer.append(
+          reinterpret_cast<const char*>(&high_part), sizeof(int64_t));
+        buffer.append(
+          reinterpret_cast<const char*>(&low_part), sizeof(uint64_t));
+
+        return serde::parquet::fixed_byte_array_value{.val = std::move(buffer)};
+    }
+};
+struct value_converting_visitor {
+    ss::future<parquet_conversion_outcome>
+    operator()(iceberg::primitive_value value) {
+        co_return std::visit(
+          primitive_value_converting_visitor{}, std::move(value));
+    }
+    ss::future<parquet_conversion_outcome>
+    operator()(std::unique_ptr<iceberg::struct_value> value) {
+        serde::parquet::group_value group;
+        group.reserve(value->fields.size());
+        for (auto& field : value->fields) {
+            if (!field.has_value()) {
+                group.emplace_back(serde::parquet::null_value{});
+                continue;
+            }
+            auto result = co_await to_parquet_value(std::move(*field));
+            if (result.has_error()) {
+                co_return result.error();
+            }
+
+            group.emplace_back(std::move(result.value()));
+        }
+        co_return group;
+    }
+    ss::future<parquet_conversion_outcome>
+    operator()(std::unique_ptr<iceberg::list_value> list) {
+        serde::parquet::repeated_value repeated;
+        repeated.reserve(list->elements.size());
+        for (auto& element : list->elements) {
+            if (!element.has_value()) {
+                repeated.emplace_back(serde::parquet::null_value{});
+                continue;
+            }
+            auto result = co_await to_parquet_value(std::move(*element));
+            if (result.has_error()) {
+                co_return result.error();
+            }
+            repeated.emplace_back(std::move(result.value()));
+        }
+        co_return repeated;
+    }
+    ss::future<parquet_conversion_outcome>
+    operator()(std::unique_ptr<iceberg::map_value> map) {
+        serde::parquet::repeated_value repeated;
+        repeated.reserve(map->kvs.size());
+        for (auto& kv : map->kvs) {
+            serde::parquet::group_value group;
+            group.reserve(2);
+            auto key_result = co_await to_parquet_value(std::move(kv.key));
+            if (key_result.has_error()) {
+                co_return key_result.error();
+            }
+            group.emplace_back(std::move(key_result.value()));
+            if (kv.val.has_value()) {
+                auto val = co_await to_parquet_value(std::move(kv.val.value()));
+                if (val.has_error()) {
+                    co_return val.error();
+                }
+                group.emplace_back(std::move(val.value()));
+            } else {
+                group.emplace_back(serde::parquet::null_value{});
+            }
+            repeated.emplace_back(std::move(group));
+        }
+        co_return repeated;
```

Comment:
These need to match the logical types like we translated in the schema - so you need to wrap them in an extra group.